Icf 167 create epi api

DookTibs opened the pull request:
draft PR as we discussed...I took a look at how animal API works and reworked a few things based on that, and had some other ideas to make it more streamlined. I still kept the whole "get or create" and the "single or list" ideas b/c I think those are pretty useful.

I am now hooking into forms for validation in e.g. OutcomeWriteSerializer; I thought on the call you had said you had broken out form validation into a separate class and the API and form both talked to that, but it lookked like in the animal api it just called the form.is_valid directly?

The custom_perm_checker/getOrCreatePermissionCheckHelper stuff I mentioned in the email the other day is also gone now; I found a better way to do that I think, passing in the user to the serializer on construction and then the serializer's (or the mixin's) validate method can use that for permission checking. Let me know what you think of that. That certainly made things less complicated than in the email I sent the other day.

So to restate those outstanding questions:
1. Does this general “get/create” approach make sense/seem useful?

2. Is this method of checking permissions (passing in user to the serializer and doing it during validate) acceptable?
 
3. When I actually check the permissions, I’m using user_can_edit_object. But with this get/create approach sometimes you aren’t actually editing, you’re just retrieving, and in those times maybe user_can_view_object is the strictly right one to call? I think it makes more sense to just always call user_can_edit_object but if you feel strongly the other way I can look at a way to do it cleaner.

shapiromatron commented:
thanks @DookTibs! I'll spend some time digging in... also included @rabstejnek for a review too...

rabstejnek commented on hawc/apps/common/api/mixins.py:
```diff
@@ -108,3 +110,113 @@ class LegacyAssessmentAdapterMixin:
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        user = kwargs.get("context", {}).get("user")
```
Since viewsets add `request` to the context by default you can access the user that way (`request.user`) instead of passing user in explicitly. That would simplify the `get_serializer` calls a little bit.

rabstejnek commented on hawc/apps/common/api/mixins.py:
```diff
@@ -108,3 +110,113 @@ class LegacyAssessmentAdapterMixin:
+    # user in question. If there is no associated assessment this
+    # will fail an assertion and the enclosing serializer must
+    # instead implement its own validation routine and not call this one.
+    def validate(self, data):
```
Maybe `data.get("assessment")` can be checked at the viewset permissions level on `POST` requests, that way we're not checking permissions both here and there. @shapiromatron what do you think?

rabstejnek reviewed (commented):
Looks good!

shapiromatron reviewed (commented):
Looks good overall, full steam ahead. I'd recommend reviewing a few patterns for the animal API to mirror some of the patterns there.

If it's a simple create with a single form (study pop, outcome), you can follow the animal.Experiment patterns:
- https://github.com/shapiromatron/hawc/blob/main/hawc/apps/animal/api.py#L178-L181
- https://github.com/shapiromatron/hawc/blob/main/hawc/apps/animal/serializers.py#L30-L50
- use some of the helper functions we created eg., get_matching_instance

If it's a form + formset (eg., result, comparison set), take a look at the animal.Endpoint pattern:
- https://github.com/shapiromatron/hawc/blob/main/hawc/apps/animal/serializers.py#L334-L400
- https://github.com/shapiromatron/hawc/blob/eed910c723bfd716b080386b5348caa347fbd659/hawc/apps/animal/forms.py#L532-L583 (though if we can just pass the serializer to the form as you did, that's even better)
- https://github.com/shapiromatron/hawc/blob/main/hawc/apps/animal/serializers.py#L234-L245

However, one pattern that you suggest that I'd rather not use is the POST that can create both a single or multiple instance of the object.  We dont use that pattern elsewhere and it's not expected REST behavior. If you want, i'd make a single creation with a POST to /{resource}/, and a multiple creation to a POST /{resource}/multiple/, it'd be easy to just make a second view I think, with less conditional checking and just expects a single instance an a list of instances, respectively.

DookTibs commented:
@shapiromatron just pushed a bunch more changes, here's the current status and notes:

* API is now functional - all epi stuff should now be creatable/updateable/deletable.

* Includes a publicly visible metadata endpoint (/epi/api/metadata/). I also added an assessment-specific version if you append an assessment id, e.g. /epi/api/metadata/508/; this requires view permission on the assessment in question and includes assessment-specific values like adjustment factors and criteria that are omitted from the public one.

* I have added pytests for all this

* I have NOT yet added support for any of this to the HawcClient

* I have NOT yet deployed this to our AWS dev instance; planning to do this over the next day or so. So I've done a lot of testing but we haven't actually started using this "for real".

* as mentioned on the last call, pretty much all the fields that are tied to choices tuples (ex: StudyPopulation.DESIGN_CHOICES) will accept either the value ("CO") or a case-insensitive readable version ("CoHoRt"). This also works for fields that are tied to other db tables; for instance StudyPopulation.countries can be specified by either the id of the row in the db table, or by the name like "Germany". This feels like a nice feature to me and there is interest in making use of this on our side, but if this feels awful to you I can always rip it out.

* for fields that are tied to another list but allow you to add to said list (for example - when adding adjustment factors to a result); these also will accept either id's or display values. If a display value the api will look up the underlying object; if it fails, it will create them on the fly. The idea here is to make it pretty analogous to the web UI which will let you add new ones, but I want to call this out b/c again I could see this feeling awful, so I can rip it out if you hate it.

* similarly, dtxsid's can be specified that are not yet in HAWC; they will be imported if possible.

* one thing I shelved at the last instance is the FormIntegrationMixin but you'll see remnants of this in the code. Basically serializers could mix this in and during a validate call I would construct a form (so like OutcomeSerializer would instantiate OutcomeForm) and call is_valid on it. I ran into a bunch of issues with this breaking some existing tests, and worse it would break during PATCH calls to the API. For example if "metric" is required for a result, and I PATCHed an existing result with just a "name" update, then the form.is_valid call would complain that metric wasn't supplied. I could probably code around this but didn't want to hold up the rest of things while I worked on that.

Anyway take a look and let me know if there are changes you'd like or if not I can submit it as a real PR. Thx!

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+
+                request.data[data_key] = fixed
+
+    def perform_create(self, serializer):
```
wrap this and all others in a transaction in case we get in a bad state and need to revert https://docs.djangoproject.com/en/3.1/topics/db/transactions/#django.db.transaction.atomic

Please do throughout the api.py

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+        super().perform_create(serializer)
+        self.process_criteria_association(serializer, serializer.data["id"], True)
+
+    def perform_update(self, serializer):
```
same re: transactions

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+
+    def get_queryset(self, *args, **kwargs):
```
move this to top since it's "standard"

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+    def process_criteria_association(self, serializer, study_population_id, post_initial_create):
+        # this should probably be rewritten to use the ManyToManyManager on the underlying instance...
+        inserts = []
+        for cc in self.criteria_categories:
```
Use tuple unpacking:
```
for (data_key, type_code) in self.criteria_cateogories:
```

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+            data_key = cc[0]
+            type_code = cc[1]
+
+            if data_key in self.request.data:
```
do in two passes? first find what we need to delete and what we need to insert? do all deleets and all inserts at once?

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+    def get_queryset(self, *args, **kwargs):
+        return self.model.objects.all()
+
+    def handle_criteria(self, request, during_create):
```
i think the serializer should have a study instance you can just use. `serializer.instance.study`; no need for more db calls

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+    serializer_class = serializers.CriteriaSerializer
+
+
+class StudyPopulation(PermCheckerMixin, viewsets.ModelViewSet):
```
why can't we use an AssessmentViewset or AssessmentEditViewset - this adds all the permission checking built-in. 

Same with other Viewsets - I'd rather not use generic DRF stuff again.

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+class Exposure(ReadWriteSerializerMixin, PermCheckerMixin, AssessmentEditViewset):
```
do we need PermChecker mixin with the AssessmentEditViewset?

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+                if "estimate_type" in ct:
+                    probe_ct_estimate_type = ct["estimate_type"]
+                    if type(probe_ct_estimate_type) is str:
+                        converted_estimate_type = find_matching_list_element_by_value(
```
it seems like we're doing it wrong; this seems too low low level

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+            if not during_update:
+                raise ValidationError(f"At least one central tendency is required")
+
+    def handle_dtxsid(self, request):
```
you are nicer than me :) nice job

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+                    metric_units = DoseUnits.objects.get(name=metric_units_probe)
+                    request.data["metric_units"] = metric_units.id
+                except ObjectDoesNotExist:
+                    # option 1 - require a pre-existing dose unit
```
use option 1

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+            ct_serializer.is_valid(raise_exception=True)
+            ct_serializer.save()
+
+    def perform_create(self, serializer):
```
tranactions

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+    model = models.GroupResult
+    serializer_class = serializers.GroupResultSerializer
+
+    # note - this does not currently validate that a supplied group is part of the assessment.
```
i think it should, yes?

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+
+    def process_adjustment_factor_creation(self, serializer, result_id, post_initial_create):
+        inserts = []
+        for fc in self.factor_categories:
```
tuple unpacking as above

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+                    inserts.append(dynamic_obj)
+
+        # ...and save any new ones
+        if len(inserts) > 0:
```
same thought as above, do bulk deletes/creates after inspecting data

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+
+                try:
+                    outcome = models.Outcome.objects.get(id=outcome_id)
+                    assessment_id = outcome.get_assessment().id
```
just use outcome.assessment_id

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
+        serializer = self.get_serializer(instance)
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+    def perform_create(self, serializer):
```
transactions

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
     model = models.ComparisonSet
     serializer_class = serializers.ComparisonSetSerializer
 
+    # note - this does not currently validate that a supplied exposure is a valid one...
```
please add

shapiromatron commented on hawc/apps/epi/api.py:
```diff
@@ -139,3 +540,21 @@ class ExposureCleanup(CleanupFieldsBaseViewSet):
+    def list(self, request):
+        return EpiMetadata.handle_request(request)
+
+    def retrieve(self, request, *args, **kwargs):
```
make this dumber; require an assessment. use mixins.RetrieveModelMixin, and just check that assessment.user_can_view_object(request.user). No list view.

shapiromatron reviewed (commented):
Lots to discuss here. I can hop on a call if it'd be beneficial.
